keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1009 @@
+
+### Vocabulary: OCI Images, Artifacts, and Objects
+
+**OCI Image ([spec](https://github.com/opencontainers/image-spec/blob/main/spec.md)):**
+   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
+     It includes a filesystem bundle and metadata required to run a container.
+   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
+     (which provides configuration data such as environment variables, entry points, etc.).
+   - **Use Case:** Used primarily for packaging and distributing containerized applications.
+
+**OCI Artifact ([guidance](https://github.com/opencontainers/image-spec/blob/main/artifacts-guidance.md)):**
+   - An artifact describes any content that is stored and distributed using the OCI image format.
+     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
+   - Artifacts use the same image manifest and layer structure but may contain different types of data
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps to determining the file location on the host filesystem.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+      pullPolicy: IfNotPresent
```

Comment:
We need to specify when the pull policy is triggered, because this is a pod level construct now and the previous pull policy was container level.  

I.e. containers repull when the container is recreated, but pullPolicy: Always on a volume spec that is mounted into two different containers is not the same as containers and needs defined semantics.

Reply:
I added more details about that point in 3f72dced